### PR TITLE
[extension/opamp]: Add mechanism to detect whether the collector has been orphaned

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:** <Describe what has changed.>
* Allows the process to monitor a passed in ppid, which should be the parent process ID for the collector. When the parent process ID exits, the extension emits a fatal error event, which triggers a collector shutdown.

**Link to tracking Issue:** This is part of #32189 - It does not resolve this issue because the supervisor still needs changes to pass the its PID in.

**Testing:**
Added some unit tests.

I've manually tested it on my macbook with this PR: https://github.com/observIQ/opentelemetry-collector-contrib/pull/4715
(running supervisor, kill -9 the supervisor, and take a look at the agent logs to see it shut down).

I've tried testing this out on Windows, but the supervisor doesn't get past bootstrapping (the Commander's Stop function does not work on windows), so I wasn't able to fully test it. 

**Documentation:** 
Added new parameter to README

### CODE REVIEW

**Review by evan-bradley — COMMENTED** [review]
Sorry, I forgot to publish my review. Mostly looks good to me, just a few comments. I'll take another look through it shortly.

File: extension/opampextension/README.md [hunk 1]
```diff
@@ -32,6 +32,8 @@ The following settings are optional:
   - `reports_effective_config`: Whether to enable the OpAMP ReportsEffectiveConfig capability. Default is `true`.
 - `agent_description`: Setting that modifies the agent description reported to the OpAMP server.
   - `non_identifying_attributes`: A map of key value pairs that will be added to the [non-identifying attributes](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#agentdescriptionnon_identifying_attributes) reported to the OpAMP server. If an attribute collides with the default non-identifying attributes that are automatically added, the ones specified here take precedence.
+- `ppid`: An optional process ID to monitor. When this process is no longer running, the extension will emit a fatal error, causing the collector to exit.
```
Comment (tigrannajaryan) [inline extension/opampextension/README.md:35]: Can we add some docs that explain that this is normally set by the Supervisor (or some other parent process) and is not something users should set manually?

File: extension/opampextension/monitor_ppid.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/shirou/gopsutil/v3/process"
+	"go.opentelemetry.io/collector/component"
+)
+
+var orphanPollInterval = 5 * time.Second
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply extension/opampextension/monitor_ppid.go]: I made this a configurable option.

I think the parameter's value less related to how often the supervisor restarts the collector, and more related to how quickly whatever is managing the supervisor restarts it (e.g. if it runs as a systemd service or windows service, it'll be restarted if it ever crashes)

File: extension/opampextension/README.md (hunk 1 above)
Comment (BinaryFissionGames) [reply extension/opampextension/README.md]: Good call, I've added this to the description for the parameter.

File: extension/opampextension/monitor_ppid_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+			t.Logf("Status event error: %s", se.Err())
+			require.FailNow(t, "status report function should not be called")
+		}
+
+		monitorCtx, monitorCtxCancel := context.WithCancel(context.Background())
+
+		setOrphanPollInterval(t, 10*time.Millisecond)
+
+		go func() {
+			time.Sleep(500 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply extension/opampextension/monitor_ppid_test.go]: I trimmed this down and marked these tests as parallel, should keep the test time more manageable.

**Review by evan-bradley — APPROVED** [review]

File: extension/opampextension/opamp_agent.go [hunk 4]
```diff
@@ -99,6 +111,10 @@ func (o *opampAgent) Start(ctx context.Context, _ component.Host) error {
 }
 
 func (o *opampAgent) Shutdown(ctx context.Context) error {
+	if o.lifetimeCtxCancel != nil {
+		o.lifetimeCtxCancel()
+	}
+
```
Comment (srikanthccv) [inline extension/opampextension/opamp_agent.go:117]: Curious if this context cancellation could result in incorrectly reporting the fatal error during the normal shutdown?

Comment (srikanthccv) [reply extension/opampextension/opamp_agent.go:117]: I just checked the return values for `process.PidExistsWithContext` for the cancelled context. It returns true and nil for running process pid so it's not an issue.

Comment (BinaryFissionGames) [reply extension/opampextension/opamp_agent.go:117]: Thanks for checking on that!

**Pull request merged**